### PR TITLE
Custom deserializers for empty string and settings type

### PR DESCRIPTION
**Pull request by StanIwan**
## GitHub issue or Phabricator ticket number?
https://phorge.wireload.net/T7350

### CODE REVIEW

File: src/commands/mod.rs [hunk 1]
```diff
@@ -226,11 +231,24 @@ pub struct EdgeAppManifest {
     pub settings: Vec<Setting>,
 }
 
+#[derive(Clone, Debug, Eq, PartialEq, Serialize)]
+pub enum EdgeAppSettingType {
```
Comment (sergey-borovkov) [inline src/commands/mod.rs:235]: Just call it Setting type

File: src/commands/mod.rs [hunk 2]
```diff
@@ -226,11 +231,24 @@ pub struct EdgeAppManifest {
...
+        EdgeAppSettingType::String
+    }
+}
+
 // maybe we can use a better name as we have EdgeAppSettings which is the same but serde_json::Value inside
 #[derive(Clone, Debug, Default, Eq, PartialEq, Serialize, Deserialize)]
 pub struct Setting {
     #[serde(rename = "type")]
-    pub type_: String,
+    #[serde(deserialize_with = "deserialize_setting_type")]
```
Comment (sergey-borovkov) [inline src/commands/mod.rs:250]: write it in one #[serde...] block with ,

File: src/commands/mod.rs [hunk 3]
```diff
@@ -268,6 +286,30 @@ where
     map.end()
 }
 
+fn deserialize_string_not_empty<'de, D>(deserializer: D) -> Result<String, D::Error>
+where
+    D: Deserializer<'de>,
+{
+    let s: String = Deserialize::deserialize(deserializer)?;
+    if s.is_empty() {
+        return Err(serde::de::Error::custom("String field must not be empty"))
```
Comment (sergey-borovkov) [inline src/commands/mod.rs:295]: will this field name?

File: .gitignore [hunk 4]
```diff
@@ -13,3 +13,4 @@ Cargo.lock
 # Added by cargo
 
 /target
+**/screenly.yml
```
Comment (sergey-borovkov) [inline .gitignore:16]: hmm?

Comment (sergey-borovkov) [reply .gitignore:16]: but there is no screenly.yml anywhere in this repo - no need I think

File: src/commands/mod.rs [hunk 5]
```diff
@@ -664,11 +728,11 @@ mod tests {
 
         let expected_contents = r#"---
 app_id: test_app
-user_version: ''
-description: ''
-icon: ''
-author: ''
-homepage_url: ''
+user_version: null
+description: null
```
Comment (sergey-borovkov) [inline src/commands/mod.rs:732]: What is this ? We don't support null. We only support strings... or ut should not be in the manifest file

**Review by sergey-borovkov — CHANGES_REQUESTED** [review]
1) needs unit tests 
2) needs to handle values being none in the payload 
3) there is no validate command
4) needs to check that app id passed to functions is not empty or show an error in that case

File: src/commands/mod.rs [hunk 6]
```diff
@@ -268,6 +285,53 @@ where
     map.end()
 }
 
+fn deserialize_string_not_empty<'de, D>(deserializer: D) -> Result<String, D::Error>
+where
+    D: Deserializer<'de>,
+{
+    let s: String = Deserialize::deserialize(deserializer)?;
+    if s.is_empty() {
+        return Err(serde::de::Error::custom("String cannot be empty"))
```
Comment (sergey-borovkov) [inline src/commands/mod.rs:294]: I asked before - will this show field name ?

**Review by sergey-borovkov — CHANGES_REQUESTED** [review]

File: src/commands/mod.rs (hunk 6 above)
Comment (StanIwan) [reply src/commands/mod.rs:294]: I've replied back then (https://github.com/Screenly/cli/pull/96#discussion_r1313016497): I don't think that you can get field name here, you get line and column.

File: src/commands/mod.rs (hunk 5 above)
Comment (StanIwan) [reply src/commands/mod.rs:732]: So if we have EdgeAppManifest like this (default for optional struct fields):

```       
 let manifest = EdgeAppManifest {
            app_id: "test_app".to_string(),
            settings: vec![Setting {
                title: "username".to_string(),
                type_: SettingType::String,
                default_value: "stranger".to_string(),
                optional: true,
                help_text: "An example of a setting that is used in index.html".to_string(),
            }],
            ..Default::default()
        };
```
It should not write this fields to the `screenly.yml`?

Comment (sergey-borovkov) [reply src/commands/mod.rs:732]: > So if we have EdgeAppManifest like this (default for optional struct fields):
> 
> ```
>  let manifest = EdgeAppManifest {
>             app_id: "test_app".to_string(),
>             settings: vec![Setting {
>                 title: "username".to_string(),
>                 type_: SettingType::String,
>                 default_value: "stranger".to_string(),
>                 optional: true,
>                 help_text: "An example of a setting that is used in index.html".to_string(),
>             }],
>             ..Default::default()
>         };
> ```
> 
> It should not write this fields to the `screenly.yml`?

If they are empty it should not.

Comment (sergey-borovkov) [reply src/commands/mod.rs:732]: Update after daily: not going to check that those Optional fields are not empty (except for app_id -> app_id should still be checked for length)

File: src/commands/edge_app.rs [hunk 7]
```diff
@@ -1242,11 +1242,11 @@ mod tests {
         let command = EdgeAppCommand::new(authentication);
         let manifest = EdgeAppManifest {
             app_id: "01H2QZ6Z8WXWNDC0KQ198XCZEW".to_string(),
-            user_version: "1".to_string(),
-            description: "asdf".to_string(),
-            icon: "asdf".to_string(),
-            author: "asdf".to_string(),
-            homepage_url: "asdfasdf".to_string(),
+            user_version: Some("1".to_string()),
```
Comment (sergey-borovkov) [inline src/commands/edge_app.rs:1245]: This is not related to this task 100%. But can you move creation of EDgeAppManifest to a separtate function for those that use the same values please?

File: src/commands/mod.rs (hunk 6 above)
Comment (sergey-borovkov) [reply src/commands/mod.rs:294]: I see, ok that's fine for now.

File: src/commands/mod.rs [hunk 8]
```diff
@@ -680,4 +909,93 @@ settings:
...
+
+        write_to_tempfile(&dir, file_name, content);
+        let file_path = dir.path().join(file_name);        
+        assert_eq!(EdgeAppManifest::validate_file(&file_path).unwrap(), false);
+    }
+
+    #[test]
+    fn test_validate_file_when_empty_field_should_return_false() {
+        let dir = tempdir().unwrap();
+        let file_name = "test.yaml";
```
Comment (sergey-borovkov) [inline src/commands/mod.rs:964]: it's screenly.yml

Comment (StanIwan) [reply src/commands/mod.rs:964]: It doesn't matter for tests but I can change it

File: src/commands/mod.rs [hunk 9]
```diff
@@ -275,6 +315,43 @@ where
...
+
+fn deserialize_setting_type<'de, D>(deserializer: D) -> Result<SettingType, D::Error>
+where
+    D: Deserializer<'de>,
+{
+    let s: String = Deserialize::deserialize(deserializer)?;
+    match s.to_lowercase().as_str() {
+        "string" => Ok(SettingType::String),
+        "secret" => Ok(SettingType::Secret),
+        _ => Err(serde::de::Error::custom("Field 'type_' must be either String or Secret")),
```
Comment (sergey-borovkov) [inline src/commands/mod.rs:351]: it should be field type...

File: src/commands/mod.rs [hunk 10]
```diff
@@ -288,6 +365,16 @@ impl EdgeAppManifest {
         write!(&manifest_file, "---\n{yaml}")?;
         Ok(())
     }
+
+    pub fn validate_file(path: &Path) -> Result<bool, CommandError> {
+        match serde_yaml::from_str::<EdgeAppManifest>(&fs::read_to_string(path)?) {
```
Comment (sergey-borovkov) [inline src/commands/mod.rs:370]: we already have code that constructs Manifest. It can be reused.

File: src/commands/mod.rs [hunk 11]
```diff
@@ -275,6 +315,43 @@ where
...
+        _ => Ok(opt),
+    }
+}
+
+
+fn serialize_setting_type<S>(setting_type: &SettingType, serializer: S) -> Result<S::Ok, S::Error>
+where
+    S: serde::Serializer,
+{
+    let s = match setting_type {
```
Comment (sergey-borovkov) [inline src/commands/mod.rs:336]: just call to_string(0) (Implement display trait or auto derive it and then convert to lower case)

File: src/commands/mod.rs (hunk 9 above)
Comment (sergey-borovkov) [reply src/commands/mod.rs:351]: Also please instead implement from str on the enum

Comment (StanIwan) [reply src/commands/mod.rs:351]: "Field type must be either String or Secret"?

File: src/cli.rs [hunk 12]
```diff
@@ -356,6 +356,12 @@ pub enum EdgeAppCommands {
         #[arg(short, long)]
         app_id: Option<String>,
     },
+    /// Validates screenly.yml manifest file
```
Comment (sergey-borovkov) [inline src/cli.rs:359]: Validates Edge App manifest file

File: src/cli.rs [hunk 13]
```diff
@@ -1130,6 +1136,22 @@ pub fn handle_cli_edge_app_command(command: &EdgeAppCommands) {
...
             edge_app_command.generate_mock_data(&manifest_path).unwrap();
+        },
+        EdgeAppCommands::Validate { path } => {
+            let manifest_path = transform_edge_app_path_to_manifest(path);
+            match EdgeAppManifest::validate_file(&manifest_path) {
+                Ok(true) => {
+                    println!("Manifest file is valid.");
+                },
+                Ok(false) => {
+                    println!("Manifest file is invalid.");
```
Comment (sergey-borovkov) [inline src/cli.rs:1147]: This is not good. More details... What is user supposed to do with this info?? You are not printing the actual error...

Comment (sergey-borovkov) [reply src/cli.rs:1147]: Why is it even returning false? If it's not validated it should error. It should be called ensure_manifest_is_validated or something

File: src/commands/mod.rs (hunk 9 above)
Comment (sergey-borovkov) [reply src/commands/mod.rs:351]: Field type should be one of the following:
- xxx
- xxx

And for that you ideally enumerate all enum values. It's ok if you don't do that programmatically. But it should be simple and won't require changes in any extra places if we add new setting type.

Comment (sergey-borovkov) [reply src/commands/mod.rs:351]: This can be done with macros of some crate like https://docs.rs/strum/latest/strum/

File: src/cli.rs (hunk 13 above)
Comment (StanIwan) [reply src/cli.rs:1147]: Printing error is inside the validate_file - I'll adjust it

Comment (sergey-borovkov) [reply src/cli.rs:1147]: But why can your function still return an error then? Makes no sense to me tbh. If you already handle it inside your function

File: src/cli.rs [hunk 14]
```diff
@@ -1130,6 +1136,18 @@ pub fn handle_cli_edge_app_command(command: &EdgeAppCommands) {
...
             edge_app_command.generate_mock_data(&manifest_path).unwrap();
+        },
+        EdgeAppCommands::Validate { path } => {
+            let manifest_path = transform_edge_app_path_to_manifest(path);
+            match EdgeAppManifest::ensure_manifest_is_validated(&manifest_path) {
+                Ok(()) => {
+                    println!("Manifest file is valid.");
+                },
+                Err(e) => {
+                    println!("{}", e);
```
Comment (sergey-borovkov) [inline src/cli.rs:1147]: you can put e inside the formatting braces

**Review by sergey-borovkov — CHANGES_REQUESTED** [review]
There are no checks for app_id not being empty everywhere.

File: src/commands/mod.rs [hunk 15]
```diff
@@ -233,11 +267,24 @@ pub struct EdgeAppManifest {
     pub settings: Vec<Setting>,
 }
 
+#[derive(Clone, Debug, Eq, PartialEq, Serialize, Default, EnumString, Display, EnumIter)]
+pub enum SettingType {
+    #[default]
+    #[strum(serialize = "String", to_string = "string")]
```
Comment (sergey-borovkov) [inline src/commands/mod.rs:273]: why is it starting with capital letter???

**Comment by sergey-borovkov** [comment]
> 1. needs unit tests
> 2. needs to handle values being none in the payload
> 3. there is no validate command
> 4. needs to check that app id passed to functions is not empty or show an error in that case

Point 2 is not implemented. Need unit tests for how payload is treated when it's None.
Point 4 is not implemented at all.

File: src/commands/mod.rs (hunk 15 above)
Comment (StanIwan) [reply src/commands/mod.rs:273]: to represent enum value name as string 1:1 but I can just make it #[strum(serialize = "string")] and it will work

Comment (sergey-borovkov) [reply src/commands/mod.rs:273]: Why do you want to represent it 1:1? Our spec says string/secret. not String/Secret. Please don't change stuff like that.

**Comment by StanIwan** [comment]
> > 1. needs unit tests
> > 2. needs to handle values being none in the payload
> > 3. there is no validate command
> > 4. needs to check that app id passed to functions is not empty or show an error in that case
> 
> Point 2 is not implemented. Need unit tests for how payload is treated when it's None. Point 4 is not implemented at all.

Point 4 adressed with this commit -> https://github.com/Screenly/cli/pull/96/commits/a778be650d71db2321c75bd354b236e733c2d4d8

**Comment by StanIwan** [comment]
> > 1. needs unit tests
> > 2. needs to handle values being none in the payload
> > 3. there is no validate command
> > 4. needs to check that app id passed to functions is not empty or show an error in that case
> 
> Point 2 is not implemented. Need unit tests for how payload is treated when it's None. Point 4 is not implemented at all.

Point 2 addressed with -> https://github.com/Screenly/cli/pull/96/commits/6becaac38607dff79a593d2432c25126dcfb7718

File: src/commands/mod.rs [hunk 16]
```diff
@@ -288,6 +378,29 @@ impl EdgeAppManifest {
         write!(&manifest_file, "---\n{yaml}")?;
         Ok(())
     }
+
+    pub fn prepare_payload(manifest: &EdgeAppManifest) -> HashMap<&str, serde_json::Value> {
```
Comment (sergey-borovkov) [inline src/commands/mod.rs:382]: This is not public function

File: src/cli.rs [hunk 17]
```diff
@@ -508,6 +514,9 @@ fn get_actual_app_id(
     path: &Option<String>,
 ) -> Result<String, CommandError> {
     match app_id {
+        Some(id) if id.is_empty() => {
+            Err(CommandError::EmptyAppId("this".to_string()))
```
Comment (sergey-borovkov) [inline src/cli.rs:518]: what's "this?"?

File: src/commands/mod.rs [hunk 18]
```diff
@@ -737,4 +1066,53 @@ settings:
 "#
         );
     }
+
+    #[test]
+    fn test_prepare_manifest_payload_includes_some_fields() {
```
Comment (sergey-borovkov) [inline src/commands/mod.rs:1071]: No. This approach is too low level. As I mentioned it should be tested in the test_upload. You should check what actual payload ends up being.

File: src/cli.rs (hunk 17 above)
Comment (sergey-borovkov) [reply src/cli.rs:518]: Hmm. Seems to be not very obviuous. So the only command that actually needs it is upload.

Because we should not pass empty app id to function. Rather when handling CLI commands we should show an early error.

Like I mentioned to you before you need to modify get_actual_app_id to return Error. It should return an error when app_id ends up being empty.

The only one that has to be outside of this is upload because it reads the whole manifest (we could move reading manifest to CLI handling but I am ok with it staying as it is for now)

Comment (StanIwan) [reply src/cli.rs:518]: > Hmm. Seems to be not very obviuous. So the only command that actually needs it is upload.
> 
> Because we should not pass empty app id to function. Rather when handling CLI commands we should show an early error.
> 
> Like I mentioned to you before you need to modify get_actual_app_id to return Error. It should return an error when app_id ends up being empty.

Well, that's exactly what I did - first case in match throws error if the `app_id` provided by the user is empty, then if `app_id` is empty in the manifest file, a separate error is being thrown. 

> 
> The only one that has to be outside of this is upload because it reads the whole manifest (we could move reading manifest to CLI handling but I am ok with it staying as it is for now)

Comment (sergey-borovkov) [reply src/cli.rs:518]: @StanIwan sorry I can't find it for other commands like promote - can you show me where they handle that?

Comment (sergey-borovkov) [reply src/cli.rs:518]: Ok I see. Does it work properly? Did you test it with all commands?

Comment (sergey-borovkov) [reply src/cli.rs:518]: Also you can remove "this" string. I don't see the point - we always know which command we call. You can always match if it's a MissingAppId error and show that.

Comment (StanIwan) [reply src/cli.rs:518]: I tested it with some, can test it with all

File: src/commands/mod.rs (hunk 18 above)
Comment (StanIwan) [reply src/commands/mod.rs:1071]: Addressed with -> https://github.com/Screenly/cli/pull/96/commits/10ec4da1b1d7a81371fe780eba68dc2c2784ea82

File: src/cli.rs (hunk 17 above)
Comment (StanIwan) [reply src/cli.rs:518]: Now, I tested it with most commands and works the same every time. Also, updated error string.

File: src/cli.rs [hunk 19]
```diff
@@ -1130,6 +1139,18 @@ pub fn handle_cli_edge_app_command(command: &EdgeAppCommands) {
...
             edge_app_command.generate_mock_data(&manifest_path).unwrap();
+        },
+        EdgeAppCommands::Validate { path } => {
+            let manifest_path = transform_edge_app_path_to_manifest(path);
+            match EdgeAppManifest::ensure_manifest_is_validated(&manifest_path) {
+                Ok(()) => {
+                    println!("Manifest file is valid.");
+                },
+                Err(e) => {
+                    println!("{e}");
```
Comment (sergey-borovkov) [inline src/cli.rs:1150]: eprintln for errors.

**Review by sergey-borovkov — APPROVED** [review]

File: src/cli.rs [hunk 20]
```diff
@@ -508,13 +514,16 @@ fn get_actual_app_id(
...
+        },
         Some(id) => Ok(id.clone()),
         None => {
             let manifest =
                 EdgeAppManifest::new(transform_edge_app_path_to_manifest(path).as_path()).unwrap();
             match manifest.app_id {
-                Some(id) => Ok(id),
-                None => {
+                Some(id) if !id.is_empty() => Ok(id.clone()),
+                _ =>  {
```
Comment (korvyashka) [inline src/cli.rs:526]: Minor: None is ok here and more readable - what else could it be?)

File: src/commands/edge_app.rs [hunk 21]
```diff
@@ -954,6 +950,18 @@ mod tests {
     use crate::commands::edge_app_server::MOCK_DATA_FILENAME;
     use tempfile::tempdir;
 
+    fn create_edge_app_manifest(settings: Vec<Setting>) -> EdgeAppManifest {
```
Comment (korvyashka) [inline src/commands/edge_app.rs:953]: Minor: Let's add smth to a name to point out it is specific to test.

File: src/commands/mod.rs [hunk 22]
```diff
@@ -7,9 +7,13 @@ use std::fs;
 use std::fs::File;
 use std::io::Write;
 use std::path::Path;
+use std::str::FromStr;
 use std::time::Duration;
 
 use serde::{Deserialize, Deserializer, Serialize};
+use serde_json::json;
+use strum::IntoEnumIterator;
```
Comment (korvyashka) [inline src/commands/mod.rs:15]: I don't see this one used.

File: src/commands/mod.rs [hunk 23]
```diff
@@ -275,6 +325,46 @@ where
...
+    let valid_setting_types = SettingType::iter()
+        .map(|t| t.to_string())
+        .collect::<Vec<_>>()
+        .join("\n") + "\n";
+
+    match SettingType::from_str(&s.to_lowercase()) {
+        Ok(setting_type) => Ok(setting_type),
+        Err(_) => Err(serde::de::Error::custom(format!(
+            "Field type should be one of the following:\n{}",
+            valid_setting_types
```
Comment (korvyashka) [inline src/commands/mod.rs:363]: Here it seems it will print String\nSecret instead of string\nsecret

File: src/commands/mod.rs [hunk 24]
```diff
@@ -659,16 +772,150 @@ mod tests {
...
+    title: username
+    optional: true
+    help_text: An example of a setting that is used in index.html
+"#;
+
+        assert_eq!(contents, expected_contents);
+    }
+
+    #[test]
+    fn test_save_to_file_should_skip_none_optional_fields() {
```
Comment (korvyashka) [inline src/commands/mod.rs:826]: save_manifest_to...

**Review by korvyashka — APPROVED** [review]

File: src/cli.rs (hunk 20 above)
Comment (StanIwan) [reply src/cli.rs:526]: It can be Some(id) and empty :)

File: src/commands/mod.rs (hunk 22 above)
Comment (StanIwan) [reply src/commands/mod.rs:15]: `use serde_json::json;` -> is used for ```.filter_map(|(key, value)| value.as_ref().map(|v| (*key, json!(v))))``` to omit None fields in the payload
`use strum::IntoEnumIterator;` -> is used for ```let valid_setting_types = SettingType::iter()```

File: src/commands/mod.rs (hunk 23 above)
Comment (StanIwan) [reply src/commands/mod.rs:363]: It fortunately won't because strum serialization is defined with lowercase 
```
pub enum SettingType {
    #[default]
    #[strum(serialize = "string")]
    String,
    #[strum(serialize = "secret")]
    Secret,
}
```

Comment (StanIwan) [reply src/commands/mod.rs:363]: image
